Initial pass at JWT secret backend

ekristen opened the pull request:
This implements the following: 
* JWT Secret Backend - you can create roles that implement different algorithms and configure them with secret shared keys or private keys depending on the algorithm. You then can issue tokens against a role, by providing various claims to the write method. The tokens are stored and are retrievable if you have the JIT ID after the fact.

* Docker Build capabilities - I'm a huge user of docker, and borrowed a few concepts from the `docker-machine` project, and introduced the ability to build this project using docker, can build cross platform too.

If you have docker setup, then it is simply just running `bash scripts/build-docker.sh` and it will be built cross platform. To build just for OSX `bash scripts/build-docker.sh -os="darwin" -arch="amd64"`

adamlc commented:
This looks awesome! Exactly what I was after, I was just writing a forum post and came across this, fantastic. Hopefully it gets merged :)

adamlc commented:
@ekristen just noticed something, I think "JIT" should be "JTI" according to the [spec](https://tools.ietf.org/html/rfc7519), could be a typo?

```
4.1.7.  "jti" (JWT ID) Claim

   The "jti" (JWT ID) claim provides a unique identifier for the JWT.
   The identifier value MUST be assigned in a manner that ensures that
   there is a negligible probability that the same value will be
   accidentally assigned to a different data object; if the application
   uses multiple issuers, collisions MUST be prevented among values
   produced by different issuers as well.  The "jti" claim can be used
   to prevent the JWT from being replayed.  The "jti" value is a case-
   sensitive string.  Use of this claim is OPTIONAL.
```

ekristen commented:
@adamlc fixed, thanks for the catch, it has been updated

ekristen commented:
@armon would be great to get some feedback from you and find out if we can get this merged in. Thanks!

armon commented on website/source/docs/secrets/jwt/index.html.md:
```diff
@@ -0,0 +1,277 @@
+#### POST
+
+<dl class="api">
+  <dt>Description</dt>
```
The docs here don't seem to match `jwt/config` from the examples. Copy/paste error?

ekristen replied:
Yup, I've removed this entirely in the updates I'm working on for this PR.

armon commented on website/source/docs/secrets/jwt/index.html.md:
```diff
@@ -0,0 +1,277 @@
+        Not Before: the time at which the token is not useful before. Expressed as seconds, unix time. (Default: current time)
+      </li>
+      <li>
+        <span class="param">jit</span>
```
`jit` in the docs may need to be changed to `jti`

ekristen replied:
Fixed in the upcoming update to the PR.

armon commented on builtin/logical/jwt/path_config.go:
```diff
@@ -0,0 +1,89 @@
+
+func pathConfig(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: `roles/(?P<name>\w+)/config`,
```
This path does not match the docs

ekristen replied:
This has been removed in the upcoming PR. I've merged the config in with the role creation.

armon commented:
@ekristen I guess the biggest issue for me is the split between the role definition and the configuration with the key. One is somewhat useless without the other, but they are also inseparable. Most changes to the role invalidate the key anyways (different algorithm). Why not role those together and require the key to be provided at the time of role definition?

armon commented on builtin/logical/jwt/path_issue.go:
```diff
@@ -0,0 +1,173 @@
+	})
+
+	err = req.Storage.Put(&logical.StorageEntry{
+		Key:   "tokens/" + data.Get("jti").(string),
```
Take a look at the `helper/salt` package. The `JTI` should probably be salted before being used in the path, to avoid any information leak from the backend.

armon replied:
The other issue is that these seem like the accumulate forever. Either they should have a lease attached and revoke deletes these, or we need some other way to avoid accumulation. Potentially a flag to disable retention would be useful as well.

ekristen replied:
There is a "expiration" claim for tokens, but by the spec it is optional, I could set a default for it and if someone wants to retain it forever they can pass `exp=0` ... the expiration can be associated with the lease time specifically.

armon replied:
Even if we don't set that in the response for the token itself, we should set an internal to Vault expiration so that it gets GC'd eventually.

ekristen replied:
What field do I need to provide in the internal data to ensure it gets GC'd eventually?

armon replied:
You need to implement the `Revoke` on the `Secret` to delete this and respond with a lease value. If there is a lease, Vault will track the response and invoke `Revoke` on expiration. You could also support `Renew` so that the GC of the `jti` can be deferred.

ekristen replied:
That's what I thought, please see my other comment on the main thread, advice on supporting renewals or not would be appreciated. 

> On Jul 30, 2015, at 10:41, Armon Dadgar <notifications@github.com> wrote:
> 
> In builtin/logical/jwt/path_issue.go:
> 
> > +
> > +	tokenString, err := token.SignedString([]byte(config.Key))
> > +
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	resp := b.Secret(SecretTokensType).Response(map[string]interface{}{
> > +		"jti": token.Claims["jti"].(string),
> > +		"token": tokenString,
> > +	}, map[string]interface{} {
> > +		"jti": data.Get("jti").(string),
> > +	})
> > +
> > +	err = req.Storage.Put(&logical.StorageEntry{
> > +		Key:   "tokens/" + data.Get("jti").(string),
> You need to implement the Revoke on the Secret to delete this and respond with a lease value. If there is a lease, Vault will track the response and invoke Revoke on expiration. You could also support Renew so that the GC of the jti can be deferred.
> 
> —
> Reply to this email directly or view it on GitHub.
>

ekristen commented:
@armon good point, I guess I can control not returning the key. This was my first go at golang, so I was borrowing for what others had already done. Let me fix the docs, and play with the key assignment a little and submit an update.

armon commented:
@ekristen I left some comments, this is looking great. No fundamental blockers to this feature, just some things to work through.

ekristen commented:
@armon thanks! I'll work through these and get an update pushed soon

jefferai commented:
I would urge the Docker stuff to be separated out from here, as it's totally separate functionality from the main JWT stuff and should probably be evaluated separately.

Regardless of whether that happens, my suggestion for building Vault into Docker would be to use CenturyLink's minimal containers, which produce containers with a minimum of any files in them and a minimal amount of data to move around. It is not uncommon to see Docker image sizes of a few megabytes using their builde. See https://github.com/CenturyLinkLabs/golang-builder

ekristen commented:
@armon I've been struggling with some basic concepts while fixing a few of these issues. JWT tokens are stateless and all claims are optional. Since vault is about short lived secrets, one could say that JWT tokens generated by it should be default have an expiration. 

I see two approaches.

1. Make JWT tokens more like the transit backend, and do not store any data, just allow the signing of claims via a write and respond with a token, nothing to store, no lease times, no renewals. 

2. Default with a expiration of 30 days on JWT tokens, allow the expiration to be changed by the user and store the claims used to generate the token, so with a renewal it'll re-use those claims but extend the expiration, not before and issued at times. The only problem I have with this approach is that there are use cases I have in mind where I'll be using expiration dates longer than 30 days.  A suggestion you made @armon was to disable storing.

Thoughts?

kgutwin commented:
Would configurable maximum lease durations help your use cases in approach 2? I just put in PR #469 which adds that feature; there's an ongoing discussion around what the true maximum "max lease duration" ought to be, so any feedback on your time requirements would be helpful.

ekristen commented:
I think having configurable maximums is useful. 

My use case primarily is for long expirations and I would not renew, if I needed to generate a new one I would just issue a new token.

However in keeping with the most traditional uses of JWT I think that supporting renewals and expirations is a good thing and allowing it to vary beyond 30 days provides more flexibility.

> On Jul 30, 2015, at 11:00, Karl Gutwin <notifications@github.com> wrote:
> 
> Would configurable maximum lease durations help your use cases in approach 2? I just put in PR #469 which adds that feature; there's an ongoing discussion around what the true maximum "max lease duration" ought to be, so any feedback on your time requirements would be helpful.
> 
> —
> Reply to this email directly or view it on GitHub.
>

armon commented:
@ekristen So I think it's important for us to split the expiration as set in the JWT itself vs the lifespan of the `JTI` we store. I think the expiration set on the JWT depends on the business case there and is divorced from Vault. I don't really know where JWT is used or if those are typically short or long lived.

My concern was to support the `JTI` lookups, we are currently doing an unbounded storage. We just store everything ever generated. Anything unbounded makes me flinch. I want a good story around either not storing them at all, or storing them for a fixed amount of time so that we can bound our storage and reason about the storage requirements.

jefferai commented:
@armon The other approach is to simply ensuring that the JTIs are sufficiently unique that lookups are unnecessary, at the potential expense of larger JWTs.

FWIW, the problem isn't entirely unique to this backend; the PKI backend currently also does not clean up old expired certificates, and needs to gain that facility.

Also, I hate referencing my own comment, but...https://github.com/hashicorp/vault/pull/460#issuecomment-126408592

ekristen commented:
@armon @jefferai I just updated my PR. This is what I did.

* decoupled expiration from lease (set the lease to 30 days as it is the maximum)
* added a few of the defined but optional JWT claims to the role to have defaults
* added the ability to disable leases, therefore preventing vault from storing any information, but still having the ability to securely store the signing key and sign tokens.
* updated documentation

jefferai commented:
@ekristen What you describe here: "added the ability to disable leases, therefore preventing vault from storing any information, but still having the ability to securely store the signing key and sign tokens."

I'm a bit confused as to why the signing key/tokens would be coupled to leases in the first place. Those should probably be a configured property of the backend. Leases should correspond to the length of time the JWT is valid, e.g. expiration (so I'm also not clear why you decoupled those two values, can you explain it?)

ekristen commented:
The signing key is never tied to the lease, I was just pointing out the fact that you can still sign without a lease, like the `transit` backend. 

I agree when I started out it made sense to couple the JWT token's expiration claim to the lease, unfortunately with the forced 30 day maximum lease that prevents one of my specific use cases from being able to work.

If we can get the other PR merged that allows maximum lease times to be configurable I'd have no problem linking the two back together. 

Also by disabling "leases" on a role, that role then acts like the `transit` backend, signs and returns the signed token with whatever claims were passed to it, it will not store the token or the claims.

I just want the feature merged so I can start using it for a project at work :)

ekristen commented:
@armon @sethvargo would like some feedback from you two, I would really like to get this merged. What do you think? What do you need from me to complete this?

ekristen commented:
Perhaps I should allow for two different methods. 

1. Lease is attached to the expiration claim
2. No leases, backend acts like the transit backend, doesn't store anything, no renewals, no leases.

Thoughts?

ekristen commented:
Yeah I realized that after I sent it, my apologies, but figure it best not to delete comments you are going to get emails about. In my mind I was thinking that this needed to get merged in before I could start using the functionality, but I realized after the last message that I'll just build and use my fork until such time you have enough time to review and we get things sorted out the best way to get it merged. Apologies for the additional comments.

ekristen commented:
I went ahead and created another branch that ties leases back to the expiration, but still provides a way to disable leases all together, which means it'll act like `transit` and sign the token and not store anything.

https://github.com/ekristen/vault/tree/jwt-lease

armon commented on builtin/logical/jwt/path_roles.go:
```diff
@@ -0,0 +1,197 @@
+				Description: "Private Key (RSA or EC) or String for HMAC Algorithm",
+			},
+			"lease": &framework.FieldSchema{
+				Type:        framework.TypeBool,
```
I'd make this a time duration instead of bool. Since with global default lease, and soon per-backend default lease it may not always be 30 days. Also you may only want a 24h or 3h lease, etc, but something less than the max.

armon commented on builtin/logical/jwt/path_roles.go:
```diff
@@ -0,0 +1,197 @@
+
+	if strings.HasPrefix(alg, "RS") {
+		// need RSA Private Key
+		if strings.Contains(key, "RSA PRIVATE KEY") == false {
```
I would attempt to actually parse the key, since it may still be invalid. Same for HS and EC.

armon commented on builtin/logical/jwt/secret_tokens.go:
```diff
@@ -0,0 +1,159 @@
+			},
+		},
+
+		DefaultDuration: 720 * time.Hour,
```
I'd move this into a `const` also this seems like a strange default. Maybe 24h instead?

armon commented on builtin/logical/jwt/path_issue.go:
```diff
@@ -0,0 +1,133 @@
+				Type: framework.TypeString,
+				Description: "The desired role with configuration for this request",
+			},
+			"iss": &framework.FieldSchema{
```
We should be explicit with these fields. "iss" should be "issuer", "sub" should be "subject" etc. I'd prefer clarity over saving a few bytes.

armon commented on builtin/logical/jwt/path_roles.go:
```diff
@@ -0,0 +1,197 @@
+				Default:     true,
+				Description: "Use leases, if true, tokens will be kept for 30 days, if false, tokens and claims are not stored",
+			},
+			"iss": &framework.FieldSchema{
```
Same here, "iss" to "default_issuer"

ekristen commented:
Thanks @armon for the comments, I can definitely address these.

Beyond these changes there seems to be some question as to whether or not the lease should be tied to the expiration field. Thoughts? It makes sense given the purposes of vault, but at the same time I have a use case where I just want to be able to set all the claims including the expiration how I want and just use vault to securely store the signing key and sign the token.

I have another branch where the lease and expiration is tied back together here https://github.com/ekristen/vault/tree/jwt-lease. Thoughts?

armon commented on builtin/logical/jwt/secret_tokens.go:
```diff
@@ -0,0 +1,159 @@
+		return nil, err
+	}
+
+	if role.Lease == true {
```
I think this can be refactored a little. The "data" response should always be the same in either mode. The 'role' is not important to the data part, but is useful as internal data for renew.

armon commented on builtin/logical/jwt/secret_tokens.go:
```diff
@@ -0,0 +1,159 @@
+			"claims": token.Claims,
+		})
+
+		resp.Secret.Lease = 720 * time.Hour
```
This should use the role specific lease value

armon commented on builtin/logical/jwt/secret_tokens.go:
```diff
@@ -0,0 +1,159 @@
+			return nil, err
+		}
+
+		err = req.Storage.Put(&logical.StorageEntry{
```
Why do we store this twice?

armon commented on builtin/logical/jwt/secret_tokens.go:
```diff
@@ -0,0 +1,159 @@
+		return nil, err
+	}
+
+	tresp, err := b.secretTokensCreate(req, d, claims, roleName.(string))
```
I don't think the token should be re-generated, extending the lease is enough

armon commented:
@ekristen This is getting way closer. I'd remove the docker stuff into a separate PR, as its a separate concern. It's also much easier for reviewers not to add the Godeps unless right before merge for next time.

armon commented:
@ekristen The expires claim is orthogonal to Vault leasing I think. You may want that claim to be valid for months or years, but the concern with the Vault lease is how long we store the JTI for.

ekristen commented:
The JTI is just a unique field in the JWT token, and current the backend uses it as a key to store the JWT token in the backend.

I guess I'm having a hard time understanding how JWT should be implemented in terms of the goal of this project. My understanding is to have backends that can issue secrets and expire them out after a lease period.

I would agree that there are use cases where I would want the JWT token validity to extend months or years in those cases I would see disabling lease and acting like the `transit` backend to be useful.

However I also see use cases where people would want to use JWT for auth tokens between web services where the JWT expiration would only be minutes or hours and could be renewed, in this case the token's exp field would have to be updated upon a renew and resigned and delivered. 

Thanks.

armon commented:
@ekristen What if we simplified this by decoupling the concerns and making this backend more like `transit`. The generated tokens should not be stored, and there should be no leases involved. We simply generate the tokens and return it to the client. They are free to index by `JTI` and store in redis/postgresql etc. Does that make sense?

ekristen commented:
That would simplify things in the short term, just concerned we'd be selling the feature short. :) 

Perhaps it would be best to go the simplified route like `transit` and open a ticket to gather interest and the best way to implement a lease/renew feature tied to the expiration claim of the token?

armon commented:
@ekristen That seems like the best approach. Ultimately, Vault isn't meant to be a scalable distributed data store, so anything that starts looking like "please store all this data for me" is out of scope. That seems to be what we are running into with the `JTI` stuff instead of anything security related. It is totally fair for us to claim its the applications responsibility to store those and not Vaults job.

ekristen commented:
Alright I'll work on that. I definitely see potential use for short lived JWT tokens for service authentication, that can be renewed just like an AWS token can be repeatedly. I agree it shouldn't be a store all this data for me type of setup.

ekristen commented:
I simplified it as discussed, updated documentation, removed docker components from it (put those into another PR already https://github.com/hashicorp/vault/pull/497 and I squashed all the commits down into one.

armon commented:
@ekristen Looks great! Only thing missing is some acceptance tests. There is a framework that can be used to do that, take a look at `transit` backend_test.go

ekristen commented:
Tests are in, I'll rebase with master and update the branch.

armon commented:
@ekristen Thanks! Flew under my radar for a bit
